### ISSUE TITLE
Connect cognito user with AppSync type

### ISSUE THREAD

**Issue by punksta** [aws-amplify/amplify-js#3093]
** Which Category is your question related to? **
how to? 
** What AWS Services are you utilizing? **
amplify, react-native, AppSync, Auth
** Provide additional details e.g. code snippets **
I am trying to build an app where a user can be added to friends by other user. Users should have name, avatar(image on s3), email and list of friends. Also users should be able to fetch its list of friends and see avatars and names. 
```
type User @model {
  id: ID!
  name: String!
  avatar: String
  friends: [User] @connection(name: "friends")
}
```

So my questions:
- what is proper and secure way of making a 1 to 1 relation between a appsync user and a cognito user?
- Maybe it can be done without appsync `User`? I can store just friend ids. If so, how can get public data of user by their id?

I've found this in samples,
https://github.com/aws-samples/aws-appsync-chat/blob/f1bc066fe28e95995db0492ecc3441de61561e2d/src/mobx/UserStore.js#L20-L24
that's not a secure way, because any cognito user can create appsync user record with any "username".

**Comment by jkeys-ecg-nmsu** [comment]
@punksta you can use the token's `sub` field as the ID for your User field to enforce a 1-1 constraint.

+1 because I would like to see this type of functionality built-in (what application with authenticated users won't have a User model in its schema)?

**Comment by AndrewMorsillo** [comment]
@jkeys-ecg-nmsu Can you provide a quick example of how to do that? I'm just getting started with amplify and am also curious about how to manage user model data. Would I just create a user object immediately after signup? Can you point me to documentation about the token and sub field? Thanks!

**Comment by jkeys-ecg-nmsu** [comment]
@AndrewMorsillo this documentation shows how you can get the payload (including the sub and username claims) from the ID token. https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html#amazon-cognito-user-pools-using-the-id-token

You would probably use an admin Cognito user to create your user in a Lambda trigger, which would require setting the `owner` field manually (otherwise your admin user will be the owner of the record). You can do that in the migrate users trigger or pre-token generation triggers, I believe. 

https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-migrate-user.html

**Comment by jkeys-ecg-nmsu** [comment]
@haverchuck maybe there is space for an `@user` directive to specify that a model represents a user, then automatically configure resolvers to create the model with the ID! field as the cognito:username or sub?

**Comment by AndrewMorsillo** [comment]
Is it even necessary to do that? Would it be sufficient to do the following:

```
type User @model @auth(rules: [{ allow: owner }]) {
    id: ID!
    avatarUrl: String
   ....etc
}
```

If I understand correctly the auth directive will restrict that document to its Cognito owner, no custom triggers necessary.

**Comment by jkeys-ecg-nmsu** [comment]
Yeah but it won't enforce the constraint that the `id` field must be their cognito:username or cognito:sub, so you'll need to make sure you enforce that constraint manually. Also subscriptions are not yet protected by `@auth` directive, so if you need to subscribe to user events you may need to write your own resolvers. 

https://github.com/aws-amplify/amplify-cli/issues/1043

**Comment by jkeys-ecg-nmsu** [comment]
@AndrewMorsillo please close this issue if your OP was answered though, I don't think my proposal deserves its own issue and it sounds like your issue is resolved.

**Comment by kwhitejr** [comment]
@AndrewMorsillo If you figure that out, please post back here. I've seen a lot of docs suggest the `Cognito` --> `Lambda` --> `AppSync` --> `DynamoDb` work flow to build out a Users db from Cognito events, but I haven't found any complete examples, i.e. with full CloudFormation / SAM reference architecture. Indeed, to the best of my understanding, SAM does not support Cognito as a triggering event ([docs](https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#event-source-types)).

**Comment by jkeys-ecg-nmsu** [comment]
@kwhitejr it helps if your primary key can be stored in the token, which is why this issue is actually a good solution IMO (since username/sub is not useful to an attacker and is available wherever the token is). 

I am unfamiliar with SAML and can't comment on that, might be worth its own issue though, if only to request documentation. @haverchuck

**Comment by kwhitejr** [comment]
> +1 because I would like to see this type of functionality built-in (what application with authenticated users won't have a User model in its schema)?

@jkeys-ecg-nmsu Your main point here is on the button. I'm still learning Cognito, but I also concur that the `sub` field seems best suited to map to an AppSync schema `id`.

For example
```
type User @model { // createUser mutation handled by Lambda in response to Cognito events
  userId: ID! # cognito.identity.sub
  username: String
  posts: [Post] @connection(name: "UserPosts")
}

type Post @model { // createPost mutation comes from authorized client
  authorId: ID! # User.userId
  slug: String # unrelated issue, but authorId+slug is the preferred Primary Key
  author: User! @connection(name: "UserPosts")
  text: String
}
```

**Comment by AndrewMorsillo** [comment]
Yikes! Subscriptions, connections, and search aren't protected by the auth directive? That really limits its usefulness. It seems everywhere I look there are undocumented landmines (batch invoke limit of 5 I found out about most recently) in amplify waiting to bite us if we choose to use it for our project...
